File: 39_combination_sum/Solution39.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

class Solution {
    private void backtrack(Set<List<Integer>> output, int[] candidates, int target, int currSum,
            List<Integer> currList) {
        if (currSum > target) {
            return;
        }
        if (currSum == target) {
            List<Integer> list = new ArrayList<>(currList);
            Collections.sort(list);
            output.add(list);
            return;
        }
        for (int candidate : candidates) {
            currList.add(candidate);
            backtrack(output, candidates, target, currSum + candidate, currList);
            currList.remove(currList.size() - 1);
        }
    }

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        Set<List<Integer>> output = new HashSet<>();
        backtrack(output, candidates, target, target, new ArrayList<>());
        return new ArrayList<>(output);
    }
}

File: problems/0039-combination-sum/Solution39.py
from typing import List


class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        n = len(candidates)
        combinations = []

        def dfs(i, sum, sequence):
            print(f"{sequence}: {sum}")
            if sum > target:
                return

            if sum == target:
                combinations.append(sequence[:])
                return

            if i == n:
                return

            sequence.append(candidates[i])
            dfs(i, sum + candidates[i], sequence)
            sequence.pop()

            dfs(i + 1, sum, sequence)

        dfs(0, 0, [])
        return combinations


s = Solution()
print(s.combinationSum([2, 3, 6, 7], 7))
